pkg/spanconfig/spanconfigstore/spanconfigstore.go

```diff
@@ -0,0 +1,447 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package spanconfigstore
+
+import (
+	"context"
+	"sort"
+
+	"github.com/cockroachdb/cockroach/pkg/keys"
+	"github.com/cockroachdb/cockroach/pkg/roachpb"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/util/interval"
+	"github.com/cockroachdb/cockroach/pkg/util/iterutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/errors"
+)
+
+// spanConfigStore is an in-memory data structure to store and retrieve
+// SpanConfigs associated with a single span. Internally it makes use of an
+// interval tree to store non-overlapping span configurations. It isn't safe for
+// concurrent use.
+type spanConfigStore struct {
+	tree    interval.Tree
+	idAlloc int64
+}
+
+// newSpanConfigStore constructs and returns a new spanConfigStore.
+func newSpanConfigStore() *spanConfigStore {
+	s := &spanConfigStore{}
+	s.tree = interval.NewTree(interval.ExclusiveOverlapper)
+	return s
+}
+
+// copy returns a copy of the spanConfigStore.
+func (s *spanConfigStore) copy(ctx context.Context) *spanConfigStore {
```

Comment:
I'm not going to hold up the merge over it, but I don't think this commit is a positive change. Thanks for separating it out into a stand-alone commit -- I think we should drop it. We have three variants now, and one in the future.

- spanconfig.Store;
- spanconfigstore.Store;
- spanconfigstore.spanConfigStore;
- (in the future) spanconfigstore.systemSpanConfigStore

This kind of spread, if anything, makes it difficult to talk about things meaninfully. Looking at the tests, I don't think the separation is also helping much. I understand the intent here was to make room for the stand-alone spanconfigstore.systemSpanConfigStore, but that too I don't think needs to be a wholly distinct type when instead it could be a set of fields on spanconfigstore.Store and appropriately named methods. The interface we care about for all the three types are spanconfig.StoreWriter (with it's diff-ing contract), primarily, and I think it's possible to test each component by interfacing through just the StoreWriter interface. Want to test spanconfigstore.systemSpanConfigStore? Just populate using the right kind of span.Targets. What to test spanconfigstore.systemSpanConfigStore? Do the same thing. Want to test the top-level spanconfigstore.Store? Again, just do the same thing.

Generally IMO when it's possible to test everything we care about through the high-level interface (which is the case here), I think we should. Peeking into the internals instead makes for test code that's going to be subject to more change over time than necessary.